cpp/StatementPutGet.cpp

```diff
@@ -123,3 +170,124 @@ Util::Proxy* StatementPutGet::get_proxy()
     return &m_proxy;
   }
 }
+
+bool StatementPutGet::http_put(std::string const& url,
+                               std::vector<std::string> const& headers,
+                               std::basic_iostream<char>& payload,
+                               size_t payloadLen,
+                               std::string& responseHeaders)
+{
+  if (!m_stmt || !m_stmt->connection)
+  {
+    return false;
+  }
+  SF_CONNECT* sf = m_stmt->connection;
+  void* curl_desc = get_curl_desc_from_pool(url.c_str(), sf->proxy, sf->no_proxy);
+  CURL* curl = get_curl_from_desc(curl_desc);
+  if (!curl)
+  {
+    return false;
+  }
+
+  char* urlbuf = (char*)SF_CALLOC(1, url.length() + 1);
+  sf_strcpy(urlbuf, url.length() + 1, url.c_str());
+
+  SF_HEADER reqHeaders;
+  reqHeaders.header = NULL;
+  for (auto itr = headers.begin(); itr != headers.end(); itr++)
+  {
+    reqHeaders.header = curl_slist_append(reqHeaders.header, itr->c_str());
+  }
+
+  PUT_PAYLOAD putPayload;
+  putPayload.buffer = &payload;
+  putPayload.length = payloadLen;
+  putPayload.read_callback = file_put_read_callback;
+
+  char* respHeaders = NULL;
+  sf_bool success = SF_BOOLEAN_FALSE;
+
+  success = http_perform(curl, PUT_REQUEST_TYPE, urlbuf, &reqHeaders, NULL, &putPayload, NULL,
+                         NULL, &respHeaders, get_retry_timeout(sf),
+                         SF_BOOLEAN_FALSE, &m_stmt->error, sf->insecure_mode,
+                         sf->retry_on_curle_couldnt_connect_count,
+                         0, sf->retry_count, NULL, NULL, NULL, SF_BOOLEAN_FALSE,
+                         sf->proxy, sf->no_proxy, SF_BOOLEAN_FALSE, SF_BOOLEAN_FALSE);
+
+  free_curl_desc(curl_desc);
+  SF_FREE(urlbuf);
+  curl_slist_free_all(reqHeaders.header);
+  if (respHeaders)
+  {
+    responseHeaders = respHeaders;
+    SF_FREE(respHeaders);
```

Comment:
should we free response headers when we pass them as a out parameter?

Reply:
responseHeaders is a std::string so basically here we are doing a copy. Changed the code though to make it clear.